x-pack/plugins/secops/public/components/timeline/index.tsx

```diff
@@ -4,82 +4,143 @@
+import { State } from '../../store/reducer';
+import { timelineByIdSelector } from '../../store/selectors';
 import { ColumnHeader } from './body/column_headers/column_header';
-import { ColumnRenderer } from './body/renderers';
-import { RowRenderer } from './body/renderers';
+import { Range } from './body/column_headers/range_picker/ranges';
+import { columnRenderers, rowRenderers } from './body/renderers';
 import { Sort } from './body/sort';
 import { DataProvider } from './data_providers/data_provider';
-import { OnColumnSorted, OnDataProviderRemoved, OnFilterChange, OnRangeSelected } from './events';
-import { TimelineHeader } from './header/timeline_header';
+import { ECS } from './ecs';
+import { OnColumnSorted, OnDataProviderRemoved, OnRangeSelected } from './events';
+import { Timeline } from './timeline';
 
-interface Props {
-  columnHeaders: ColumnHeader[];
-  columnRenderers: ColumnRenderer[];
-  data: ECS[];
+export interface OwnProps {
+  id: string;
+  headers: ColumnHeader[];
+  width: number;
+}
+
+interface StateProps {
   dataProviders: DataProvider[];
-  height?: string;
-  onColumnSorted: OnColumnSorted;
-  onDataProviderRemoved: OnDataProviderRemoved;
-  onFilterChange: OnFilterChange;
-  onRangeSelected: OnRangeSelected;
-  rowRenderers: RowRenderer[];
+  data: ECS[];
+  range: Range;
   sort: Sort;
-  width: number;
 }
 
-const TimelineDiv = styled.div<{ width: string; height: string }>`
-  display: flex;
-  flex-direction: column;
-  min-height: 700px;
-  overflow: none;
-  user-select: none;
-  width: ${props => props.width};
-  height: ${props => props.height};
-`;
-
-const defaultHeight = '100%';
-
-/** The parent Timeline component */
-export const Timeline = pure<Props>(
-  ({
-    columnHeaders,
-    columnRenderers,
-    dataProviders,
-    data,
-    height = defaultHeight,
-    onColumnSorted,
-    onDataProviderRemoved,
-    onFilterChange,
-    onRangeSelected,
-    rowRenderers,
-    sort,
-    width,
-  }) => (
-    <TimelineDiv data-test-subj="timeline" width={`${width}px`} height={height}>
-      <TimelineHeader
-        dataProviders={dataProviders}
-        onDataProviderRemoved={onDataProviderRemoved}
-        width={width}
-      />
-      <Body
-        columnHeaders={columnHeaders}
+interface DispatchProps {
+  createTimeline: ActionCreator<{ id: string }>;
+  addProvider: ActionCreator<{
+    id: string;
+    provider: DataProvider;
+  }>;
+  updateData: ActionCreator<{
+    id: string;
+    data: ECS[];
+  }>;
+  updateProviders: ActionCreator<{
+    id: string;
+    providers: DataProvider[];
+  }>;
+  updateRange: ActionCreator<{
+    id: string;
+    range: Range;
+  }>;
+  updateSort: ActionCreator<{
+    id: string;
+    sort: Sort;
+  }>;
+  removeProvider: ActionCreator<{
+    id: string;
+    providerId: string;
+  }>;
+}
+
+type Props = OwnProps & StateProps & DispatchProps;
+
+class StatefulTimelineComponent extends React.PureComponent<Props> {
```

Comment:
I think this is easy to read and pretty acceptable code. However, I do want to point out per our discussions and what is going on with React hooks and React's commitment to composition over classes that it might be better to start PR's using `lifecycle` from recompose to get in good habits when needed things like `componentDidMount`:
https://github.com/acdlite/recompose/blob/master/docs/API.md#lifecycle

Example:
https://github.com/acdlite/recompose/wiki/Recipes#show-a-spinner-while-a-component-is-loading

Disclaimer, I haven't used lifecyle before, but man does it look so much more composition-y then class-y.

No changes requested explicitly for this PR, but we should probably start thinking about composition more with recompose and then swap those out for the lifecycle hooks when those become available. And it might make it easier to not do the wrong thing in future with any new people.

Reply:
👍